binaries/autotuner.cc

```diff
@@ -0,0 +1,550 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+
+#include "caffe2/core/init.h"
+#include "caffe2/core/logging.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/core/timer.h"
+#include "caffe2/core/types.h"
+#include "caffe2/proto/caffe2.pb.h"
+#include "caffe2/utils/proto_utils.h"
+#include "caffe2/utils/string_utils.h"
+
+CAFFE2_DEFINE_string(net, "", "The given net to benchmark.");
+CAFFE2_DEFINE_string(
+    optimized_net,
+    "",
+    "Output filename for optimized NetDef protobuf.");
+CAFFE2_DEFINE_string(
+    init_net,
+    "",
+    "The given net to initialize any parameters.");
+CAFFE2_DEFINE_string(
+    input,
+    "",
+    "Input that is needed for running the network. If "
+    "multiple input needed, use comma separated string.");
+CAFFE2_DEFINE_string(
+    input_file,
+    "",
+    "Input file that contain the serialized protobuf for "
+    "the input blobs. If multiple input needed, use comma "
+    "separated string. Must have the same number of items "
+    "as input does.");
+CAFFE2_DEFINE_string(
+    input_dims,
+    "",
+    "Alternate to input_files, if all inputs are simple "
+    "float TensorCPUs, specify the dimension using comma "
+    "separated numbers. If multiple input needed, use "
+    "semicolon to separate the dimension of different "
+    "tensors.");
+CAFFE2_DEFINE_string(input_type, "", "Input type (uint8_t/float)");
+CAFFE2_DEFINE_bool(
+    try_winograd_fp16,
+    true,
+    "Consider WINOGRAD_FP16 NNPACK algorithm in auto-tuning.");
+CAFFE2_DEFINE_int(
+    samples,
+    7,
+    "The number of samples of warm-up & main iterations to run.");
+CAFFE2_DEFINE_int(warmup, 5, "The number of warm-up iterations to run.");
+CAFFE2_DEFINE_int(iter, 15, "The number of iterations to run.");
+
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+struct InputDescription {
+  InputDescription(const string& name, const caffe2::BlobProto& blob)
+      : name(name), blob(blob) {}
+
+  InputDescription(const string& name, caffe2::BlobProto&& blob)
+      : name(name), blob(blob) {}
+
+  InputDescription(const string& name, const vector<int> shape, bool is_float)
+      : name(name), shape(shape), is_float(is_float) {}
+
+  string name;
+  /*
+   * Protobuf with blob data and metadata.
+   * If this value is not initialized, a new blob is created based on shape and
+   * is_float values.
+   */
+  caffe2::BlobProto blob;
+  /* Shape of the input (ignored if blob is initialized). */
+  vector<int> shape;
+  /* If true, input is of float type. Otherwise, it is of uint8_t type. */
```

Comment:
caffe2 has a whole TypeMeta class that would be better to use (as this may eventually evolve in fp16 for GPU autotuning)